### ISSUE TITLE
Reflexive many-to-many associations (confusing/outdated info on StackOverflow)

### ISSUE THREAD

**Issue by mazyvan** [balderdashy/waterline#1405]
**Waterline version**: 0.11.0
**Node version**: 4.4.3
**NPM version**: 3.10.6
**Operating system**: Windows

<hr/>

Hi
I thought sails/waterline has a solution for this, but I'm afraid it doesn't http://stackoverflow.com/questions/26228071/sails-js-same-model-many-to-many-association
I think this is so much required

**Comment by sailsbot** [comment]
Hi @mazyvan!  It looks like you missed a step or two when you created your issue.  Please edit your comment (use the pencil icon at the top-right corner of the comment box) and fix the following:

* Provide your Waterline version
* Provide your Node version
* Provide your NPM version
* Provide your Operating system
* Verify "I am experiencing a concrete technical issue with Waterline (ideas and feature proposals should follow the guide for proposing features and enhancements (http://bit.ly/sails-feature-guide), which involves making a pull request).  If you're not 100% certain whether it's a bug or not, that's okay--you may continue.  The worst that can happen is that the issue will be closed and we'll point you in the right direction."
* Verify "I am not asking a question about how to use Waterline or about whether or not Waterline has a certain feature (please refer to the documentation(http://sailsjs.org), or post on http://stackoverflow.com, our Google Group (http://bit.ly/sails-google-group) or our live chat (https://gitter.im/balderdashy/sails)."
* Verify "I have already searched for related issues, and found none open (if you found a related _closed_ issue, please link to it in your post)."
* Verify "My issue title is concise, on-topic and polite."
* Verify "**I can provide steps to reproduce this issue that others can follow.**"

As soon as those items are rectified, post a new comment (e.g. &ldquo;Ok, fixed!&rdquo;) below and we'll take a look.  Thanks!

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@treeline.io.*

**Comment by mazyvan** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
Sorry to be a hassle, but it looks like your issue is still missing some required info.  Please double-check your initial comment and try again.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@treeline.io.*

**Comment by mazyvan** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
Sorry to be a hassle, but it looks like your issue is still missing some required info.  Please double-check your initial comment and try again.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@treeline.io.*

**Comment by mazyvan** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
@mazyvan,@sailsbot: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by mikermcneil** [comment]
@mazyvan Thanks for posting and apologies for not responding sooner-- there's been a lot going on.  We saw this last month, and Irl (@irlnathan) manually tested many different permutations of .populate() across reflexive associations.  While it was a harsh reminder that we really could use some automated tests for reflexive associations, fortunately, it also turned up that all of the ones he tested worked as expected.

That said, it's totally possible that we missed something, so if you're seeing something we didn't cover below, would you let us know?

Here are the test cases:

---------------------

### To test

1. please create records for Irl, Shannon, Lisa, Ronnie, Irl, Scott, Cody, Mike, and Rachael.  Link them up with each other in different ways.
2. test all of the queries below on both sails-disk
3. repeat with sails-mysql
4. repeat with sails-mysql again, this time specifying a custom column name for the primary key, for `mom`, and for `dad`.

_they should all work as expected_


### Model def

```javascript
// api/models/User.js
module.exports = {
  attributes: {
    name: { type: 'string' },
    dad: { model: 'User' },
    mom: { model: 'User' },
    dadOfChildren: { collection: 'User', via: 'dad' },
    followers: { collection: 'User', via: 'following' },
    following: { collection: 'User', via: 'followers' },
    friends: { collection: 'User' }
  }
}
```



### Queries

```javascript
// (A) Sanity
User.findOne({ name: 'irl' }).exec(...);
User.findOne({ name: 'ronnie' }).exec(...);

// (B) Basic tests
User.findOne({ name: 'ronnie' }).populate('dad').exec(...);
User.findOne({ name: 'ronnie' }).populate('mom').exec(...);
User.findOne({ name: 'ronnie' }).populate('dadOfChildren').exec(...); // (..hopefully... none..!)
User.findOne({ name: 'ronnie' }).populate('followers').exec(...);
User.findOne({ name: 'ronnie' }).populate('following').exec(...);
User.findOne({ name: 'ronnie' }).populate('friends').exec(...);
User.findOne({ name: 'irl' }).populate('dadOfChildren').exec(...);
User.findOne({ name: 'irl' }).populate('followers').exec(...);
User.findOne({ name: 'irl' }).populate('following').exec(...);
User.findOne({ name: 'irl' }).populate('friends').exec(...);

// (C) Then:
// Repeat all "Basic tests" using only a `where` in the subcriteria
// (for example `.populate('friends', { where: { age: {'<': 20} } })`)

// (D) And then:
// Repeat all "Basic tests" using only a `sort`+`limit` in the subcriteria
// (for example `.populate('friends', { sort:'name DESC', limit: 1 })`)

// (E) And finally:
// Repeat all "Basic tests" using a `sort`+`limit`+`where` in the subcriteria
// (for example `.populate('friends', { where: { age: {'<': 20} } }, sort:'name DESC', limit: 1 })`)

```


I went ahead and flagged this issue to keep it open in the mean time.

**Comment by sailsbot** [comment]
@mazyvan,@sailsbot,@mikermcneil: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by mazyvan** [comment]
hehe, thanks @sailsbot for remind me this.
@mikermcneil thanks for replying 
I recently got a job and been very busy. So it was impossible for me to continue with the personal project I had.
I just checked that and It works perfectly on sails v **0.12.11** but when I opened this issue with the sails' version that I had It didn't work. I don't remember what version was.
I have no idea why.
I think this issue is closed because now it works.
